lectures/zksnark.tex

```diff
@@ -0,0 +1,812 @@
+= \langle c, w\rangle$. Where vectors $a, b, c, w \in \mathbb{F}^n $. Let's expand the inner
+products: 
+\[ \sum_{i=1}^{n} a_i w_i \times \sum_{j=1}^{n} b_j w_j = \sum_{k=1}^{n} c_k w_k \]
+
+Combine the products into a double sum on the left side:
+\[ \sum_{i=1}^{n} \sum_{j=1}^{n} a_i b_j w_i w_j = w^T (a \otimes b) w = w^T A w \]
+
+Thus, the constraint can be written as:
+\[ w^T A w + c^T w = 0 \]
+
+So, the rank-1 means the rank of the coefficients matrix $A$ in one of the constraint formats.
+
+\subsection{Quadratic Arithmetic Program}
+
+While the Rank-1 Constraint System provides a powerful way to represent computations, it is not 
+succinct at all, since the number of constraints directly depends on the complexity of the problem
+being solved. In practical scenarios, this can require tens or even hundreds of thousands of 
+constraints, sometimes even millions. The Quadratic Arithmetic Program (QAP) can address this issue.
+
+\begin{remark}
+    Understanding polynomials and their properties is crucial for this section. If you are not 
+    confident in this area, it is better to revisit the corresponding chapter and refresh your
+    knowledge.
+\end{remark}
+
+To define a constraint in the R1CS we need four vectors: three coefficient vectors ($a$, $b$, and
+$c$) and the witness one ($w$). And that's just for one constraint. As you can imagine, many of
+the values in these vectors are zeros. In circuits with thousands of inputs, outputs, and auxiliary
+variables, where there are also thousands of constraints, you could end up with millions of zeroes.
+
+So, we need to change the way how we manage coefficients. 
+
+\begin{definition} 
+    Consider a Rank-1 Constraint System (R1CS) defined by $m$ constraints. Each constraint is
+    associated with coefficient vectors $a_i b_i$ and $c_i$, where $i \in \{1, 2, \dots, m\}$ and
+    also a witness vector $w$ consisting of $n$ elements.
+
+    Then this system can also be represented using the corresponding matrices $A, B,$ and $C$.
+    \begin{align*}
+        A = \begin{pmatrix}
+            a_{11} & a_{12} & \dots & a_{1n} \\
+            a_{21} & a_{22} & \dots & a_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            a_{m1} & a_{m2} & \dots & a_{mn}
+        \end{pmatrix} & \quad
+        B = \begin{pmatrix}
+            b_{11} & b_{12} & \dots & b_{1n} \\
+            b_{21} & b_{22} & \dots & b_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            b_{m1} & b_{m2} & \dots & b_{mn}
+        \end{pmatrix} & 
+        C = \begin{pmatrix}
+            c_{11} & c_{12} & \dots & c_{1n} \\
+            c_{21} & c_{22} & \dots & c_{2n} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            c_{m1} & c_{m2} & \dots & c_{mn}
+        \end{pmatrix}
+    \end{align*}
+    
+    In this representation:
+    \begin{itemize}
+        \item Each row of the matrices corresponds to the coefficients of a specific constraint.
```

Comment:
For better understanding, I would rephrase it as "Each i-th row of the matrix corresponds to the corresponding coefficients in the i-th constraint".